### PR TITLE
Sing/feat groundlevelfacility 01

### PR DESCRIPTION
**Pull request by singyiu**
contracts:
isEntityTypeIdGroundLevelFacility
  entityTypeId <= 100

canBuildFacilityTypeAtPosition
  with the ground level facility requirements implemented as discussed

unit tested

client:
  mudPositionToEntityKey
    entityKey can be calculated from position (Vector3) with this function

  mudGetEntityMetadataAtPosition
    more mud functions are exposed in createSystemCalls
    in particular this function returns all the metadata of the facility located at input position (if the facility exists)
    example output:
```
{
  "entityKey": "0x5c6090c0461491a2941743bda5c3658bf1ea53bbd3edcde54e16205e18b45792",
  "position": {
    "x": 1,
    "y": 0,
    "z": 1
  },
  "entityTypeId": {
    "typeId": 10,
    "__staticData": "0x0000000a"
  },
  "orientation": {
    "yaw": 0,
    "__staticData": "0x00000000"
  },
  "ownedBy": {
    "owner": "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
    "__staticData": "0x70997970c51812dc3a010c7d01b50e0d17dc79c8"
  }
}
```

### CODE REVIEW

File: packages/contracts/src/systems/FacilitySystem.sol [hunk 1]
```diff
@@ -38,13 +47,35 @@ contract FacilitySystem is System {
...
+   * @param z The z coordinate of the position to check.
+   * @return True if a facility of entityTypeId can be built at the given position, false otherwise.
    */
-  function isPlayerOwnerOfFacility(address player, bytes32 entityKey) public view returns (bool) {
-    return OwnedBy.get(entityKey) == player;
+  function canBuildFacilityTypeAtPosition(uint32 entityTypeId, int32 x, int32 y, int32 z) public view returns (bool) {
+    if (y < 0) return false;
+    if (!isPositionEmpty(x, y, z)) return false;
+    if (y == 0) return isEntityTypeIdGroundLevelFacility(entityTypeId);
+    return isPositionNextToAnExistingEntity(x, y, z);
```
Comment (Cygnusfear) [inline packages/contracts/src/systems/FacilitySystem.sol:78]: Yes nice, this makes sense. 
Observing that as we grow in scope we will have to considering how to keep client logic + contract logic in sync.

File: packages/contracts/src/systems/FacilitySystem.sol [hunk 2]
```diff
@@ -79,7 +110,7 @@ contract FacilitySystem is System {
    */
   function destoryFacility(bytes32 entityKey) public {
```
Comment (Cygnusfear) [inline packages/contracts/src/systems/FacilitySystem.sol]: English spelling nit: 
`destory` should be `destroy`

File: packages/client/src/mud/createSystemCalls.ts [hunk 3]
```diff
@@ -45,22 +48,39 @@ export function createSystemCalls(
...
+
+  const mudGetOrientation = async (entityKey) => {
+    return getComponentValue(Orientation, entityKey);
+  };
+
+  const mudGetPosition = async (entityKey) => {
+    return getComponentValue(Position, entityKey);
+  };
+
+  const mudPositionToEntityKey = (pos: Vector3) => {
```
Comment (Cygnusfear) [inline packages/client/src/mud/createSystemCalls.ts:67]: forgot what keccak did 🤣 my Solidity is rusty, this makes a lot of sense.

File: packages/contracts/src/systems/FacilitySystem.sol (hunk 2 above)
Comment (singyiu) [reply packages/contracts/src/systems/FacilitySystem.sol]: Thanks. Good catch.
Fixed.

File: packages/client/src/mud/createSystemCalls.ts (hunk 3 above)
Comment (singyiu) [reply packages/client/src/mud/createSystemCalls.ts:67]: its a hash 256 function widely used in Solidity 🤣

**Pull request merged**